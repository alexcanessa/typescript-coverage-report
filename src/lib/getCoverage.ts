import {
  lint,
  AnyInfo,
  FileTypeCheckResult,
  LintOptions
} from "type-coverage-core";

type FileCount = Map<
  string,
  Pick<FileTypeCheckResult, "correctCount" | "totalCount">
>;

export type CoverageData = {
  fileCounts: FileCount;
  anys: AnyInfo[];
  percentage: number;
  total: number;
  covered: number;
  uncovered: number;
};

export type Options = Partial<
  Pick<
    LintOptions,
    "strict" | "debug" | "ignoreFiles" | "ignoreCatch" | "tsProjectFile"
  > & {
    cache: LintOptions["enableCache"];
    ignoreUnread: LintOptions["ignoreUnreadAnys"];
  }
>;

const getCoverage = async (options?: Options): Promise<CoverageData> => {
  const {
    tsProjectFile = ".",
    strict,
    debug,
    ignoreFiles,
    ignoreCatch,
    cache: enableCache,
    ignoreUnread: ignoreUnreadAnys
  } = options || {};

  const { anys, fileCounts, totalCount, correctCount } = await lint(
    tsProjectFile,
    {
      strict,
      debug,
      ignoreFiles,
      ignoreCatch,
      enableCache,
      ignoreUnreadAnys,
      fileCounts: true
    }
  );
  const percentage = totalCount === 0 ? 100 : (correctCount * 100) / totalCount;

  return {
    fileCounts,
    anys,
    percentage,
    total: totalCount,
    covered: correctCount,
    uncovered: totalCount - correctCount
  };
};

export default getCoverage;
