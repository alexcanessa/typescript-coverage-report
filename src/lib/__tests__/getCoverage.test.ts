import typeCoverageCore from "type-coverage-core";
import getCoverage from "../getCoverage";

describe("getCoverage function", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("returns calculated data from type-coverage-core", async () => {
    const data = await getCoverage();

    expect(data).toMatchSnapshot();
  });

  it("accepts a tsProjectFile option", async () => {
    typeCoverageCore.lint = jest.fn().mockImplementation(typeCoverageCore.lint);

    const data = await getCoverage({
      tsProjectFile: "./app/tsconfig.json"
    });

    expect(data).toMatchSnapshot();
    // @ts-expect-error
    expect(typeCoverageCore.lint.mock.calls).toMatchSnapshot();
  });

  it("defaults to root project when tsProjectFile is not passed", async () => {
    typeCoverageCore.lint = jest.fn().mockImplementation(typeCoverageCore.lint);

    await getCoverage();

    // @ts-expect-error
    expect(typeCoverageCore.lint.mock.calls).toMatchSnapshot();
  });
});
