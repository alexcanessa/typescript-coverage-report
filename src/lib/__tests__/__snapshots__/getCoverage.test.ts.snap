// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCoverage function accepts a tsProjectFile option 1`] = `
{
  "anys": [],
  "covered": 100,
  "fileCounts": Map {
    "index.html" => {
      "correctCount": 100,
      "totalCount": 100,
    },
  },
  "percentage": 100,
  "total": 100,
  "uncovered": 0,
}
`;

exports[`getCoverage function accepts a tsProjectFile option 2`] = `
[
  [
    "./app/tsconfig.json",
    {
      "debug": undefined,
      "enableCache": undefined,
      "fileCounts": true,
      "files": undefined,
      "ignoreCatch": undefined,
      "ignoreFiles": undefined,
      "ignoreUnreadAnys": undefined,
      "strict": undefined,
    },
  ],
]
`;

exports[`getCoverage function defaults to root project when tsProjectFile is not passed 1`] = `
[
  [
    ".",
    {
      "debug": undefined,
      "enableCache": undefined,
      "fileCounts": true,
      "files": undefined,
      "ignoreCatch": undefined,
      "ignoreFiles": undefined,
      "ignoreUnreadAnys": undefined,
      "strict": undefined,
    },
  ],
]
`;

exports[`getCoverage function returns calculated data from type-coverage-core 1`] = `
{
  "anys": [],
  "covered": 100,
  "fileCounts": Map {
    "index.html" => {
      "correctCount": 100,
      "totalCount": 100,
    },
  },
  "percentage": 100,
  "total": 100,
  "uncovered": 0,
}
`;
